### PR TITLE
e2e-varnet

### PR DESCRIPTION
**Pull request by mersad95zd**
Signed-off-by: mersad95zd <m.zalbagi@gmail.com>

Fixes # .

### Description
This pull request intends to combine ComplexUNet, CoilSensitivityModel, and VarNetBlock (all are now monai modules) to create the e2e-VarNet model.

### Status
**Work in progress**

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [ ] New tests added to cover the changes.
- [ ] Integration tests passed locally by running `./runtests.sh -f -u --net --coverage`.
- [x] Quick tests passed locally by running `./runtests.sh --quick --unittests  --disttests`.
- [x] In-line docstrings updated.
- [x] Documentation updated, tested `make html` command in the `docs/` folder.

### CODE REVIEW

**Comment by mersad95zd** [comment]
I also had model parallelism in mind to add, but will do that as an extra feature in a future PR.

**Comment by Can-Zhao** [comment]
Looks good. Could you add a unit test? Thanks!

File: monai/apps/reconstruction/networks/nets/utils.py [hunk 1]
```diff
@@ -135,6 +161,109 @@ def complex_normalize(x: Tensor) -> Sequence:  # type: ignore
...
+        raise ValueError(f"only 2D (B,C,H,W) and 3D (B,C,H,W,D) data are supported but x has shape {x.shape}")
+
+
+def divisible_pad_t(
+    x: Tensor, k: int = 16
+) -> Tuple[Tensor, Tuple[Tuple[int, int], Tuple[int, int], Tuple[int, int], int, int, int]]:  # type: ignore
+    """
+    Pad input to feed into the network (torch script compatible)
+
+    Args:
```
Comment (mersad95zd) [inline monai/apps/reconstruction/networks/nets/utils.py:174]: *New function*

Two reasons behind adding this util:
1- Monai's `DivisiblePad` and `SpatialPad` are not jit script compatible
2- There is a difference between those two and `torch.nn.functional.pad` . Don't know what's the difference exactly but somehow the latter works better in image reconstruction tasks.

File: monai/apps/reconstruction/networks/nets/utils.py [hunk 2]
```diff
@@ -135,6 +161,109 @@ def complex_normalize(x: Tensor) -> Sequence:  # type: ignore
...
+        raise ValueError(f"only 2D (B,C,H,W) and 3D (B,C,H,W,D) data are supported but x has shape {x.shape}")
+
+    return x, pad_sizes
+
+
+def inverse_divisible_pad_t(
+    x: Tensor, pad_sizes: Tuple[Tuple[int, int], Tuple[int, int], Tuple[int, int], int, int, int]
+) -> Tensor:  # type: ignore
+    """
+    De-pad network output to match its original shape
```
Comment (mersad95zd) [inline monai/apps/reconstruction/networks/nets/utils.py:232]: *New function*

Similar to divisible_pad_t.

File: monai/apps/reconstruction/networks/nets/utils.py [hunk 3]
```diff
@@ -135,6 +161,109 @@ def complex_normalize(x: Tensor) -> Sequence:  # type: ignore
...
+        return x[..., h_pad[0] : h_mult - h_pad[1], w_pad[0] : w_mult - w_pad[1], d_pad[0] : d_mult - d_pad[1]]
+
+    else:
+        raise ValueError(f"only 2D (B,C,H,W) and 3D (B,C,H,W,D) data are supported but x has shape {x.shape}")
+
+
+def floor_ceil(n: float) -> Tuple[int, int]:
+    """
+    Returns floor and ceil of the input
+
```
Comment (mersad95zd) [inline monai/apps/reconstruction/networks/nets/utils.py:256]: *New function*

File: monai/apps/reconstruction/networks/nets/varnet.py [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+from monai.apps.reconstruction.complex_utils import complex_abs_t
+from monai.apps.reconstruction.mri_utils import root_sum_of_squares_t
+from monai.apps.reconstruction.networks.blocks.varnetblock import VarNetBlock
+from monai.networks.blocks.fft_utils_t import ifftn_centered_t
+
+
+class VariationalNetworkModel(nn.Module):
+    """
+    The end-to-end variational network (or simply e2e-VarNet) based on Sriram et. al., "End-to-end variational
+    networks for accelerated MRI reconstruction".
```
Comment (mersad95zd) [inline monai/apps/reconstruction/networks/nets/varnet.py:24]: *New function*

**Comment by mersad95zd** [comment]
This PR grew in size from my initial commit, apologies for that. To make the model `jit script compatible`, I had to make adjustments to several modules that I wrote before. 

New elements are only few and I've marked those in the comments above. The rest of the codes are either minor reformatting or `Tensor` version of some functions denoted by `_t` for jit compatibility.

Please let me know if I should do anything to make this easier to review, thanks!

File: monai/apps/reconstruction/mri_utils.py [hunk 5]
```diff
@@ -33,3 +35,26 @@ def root_sum_of_squares(x: NdarrayOrTensor, spatial_dim: int) -> NdarrayOrTensor
     """
     rss_x: NdarrayOrTensor = (x**2).sum(spatial_dim) ** 0.5
     return rss_x
+
+
+def root_sum_of_squares_t(x: Tensor, spatial_dim: int) -> Tensor:
```
Comment (wyli) [inline monai/apps/reconstruction/mri_utils.py:40]: this is almost identical to `root_sum_of_squares` please reuse the existing

File: monai/apps/reconstruction/complex_utils.py [hunk 6]
```diff
@@ -144,19 +159,39 @@ def complex_mul(x: NdarrayOrTensor, y: NdarrayOrTensor) -> NdarrayOrTensor:
...
+    if x.shape[-1] != 2 or y.shape[-1] != 2:
+        raise ValueError(f"last dim must be 2, but x.shape[-1] is {x.shape[-1]} and y.shape[-1] is {y.shape[-1]}.")
+
+    real_part = x[..., 0] * y[..., 0] - x[..., 1] * y[..., 1]
+    imag_part = x[..., 0] * y[..., 1] + x[..., 1] * y[..., 0]
+
+    return torch.stack((real_part, imag_part), dim=-1)  # type: ignore
+
+
 def complex_conj(x: NdarrayOrTensor) -> NdarrayOrTensor:
```
Comment (wyli) [inline monai/apps/reconstruction/complex_utils.py:192]: I think `complex_conj` should now reuse `complex_conj_t`

File: monai/apps/reconstruction/complex_utils.py [hunk 7]
```diff
@@ -121,6 +121,21 @@ def complex_abs(x: NdarrayOrTensor) -> NdarrayOrTensor:
     return (x[..., 0] ** 2 + x[..., 1] ** 2) ** 0.5
 
 
+def complex_abs_t(x: Tensor) -> Tensor:
```
Comment (wyli) [inline monai/apps/reconstruction/complex_utils.py:124]: it's the same as `complex_abs`, let's not repeat the logic, perhaps we can remove the type annotations for now.

**Review by wyli — COMMENTED** [review]
thanks, I put some minor comments inline

**Comment by mersad95zd** [comment]
Thanks @wyli... I completely forgot about removing the redundancy between `_t` and ordinary functions... Now each function calls its script `_t` function.

**Review by wyli — APPROVED** [review]
Thanks, it looks good to me, I'll merge this PR if all tests work fine.

**Comment by mersad95zd** [comment]
Thank you!

**Pull request merged**